forward-proxy/src/main.rs

```diff
@@ -40,11 +47,52 @@ impl ProxyHttp for ForwardProxy {
         _session: &mut Session,
         _ctx: &mut Self::CTX,
     ) -> Result<Box<pingora_core::upstreams::peer::HttpPeer>> {
-        Ok(Box::from(HttpPeer::new(
-            String::from("127.0.0.1:6193"),
-            false,
-            String::from(""),
-        )))
+        // testing certs data; fixme to be dynamic
+
+        // mTLS Steps:
+        // 1. Client connects to server
+        // 2. Server presents its TLS certificate
+        // 3. Client verifies the server's certificate
+        // 4. Client presents its TLS certificate
+        // 5. Server verifies the client's certificate
+        // 6. Server grants access
+        // 7. Client and server exchange information over encrypted TLS connection
+        //
+        // Code below is for step 4(this is a client to RP), presenting the client's TLS certificate.
+        let mut peer = HttpPeer::new(
+            String::from("localhost:6193"),
+            true,
+            "localhost".to_string(),
+        );
+        {
+            let cert = X509::from_pem(include_bytes!("../../certs/generated/forward_proxy.pem"))
+                .or_err(TLS_CONF_ERR, "Failed to load CA certificate")?;
+
+            let ca_cert = X509::from_pem(include_bytes!("../../certs/generated/ca.pem"))
+                .or_err(TLS_CONF_ERR, "Failed to load CA certificate")?;
+
+            let key = boring::pkey::PKey::private_key_from_pem(include_bytes!(
+                "../../certs/generated/forward_proxy-key.pem"
+            ))
+            .or_err(TLS_CONF_ERR, "Failed to load private key")?;
+
+            // The certificate to present in mTLS connections to upstream
+            // The organization implementing mTLS acts as its own certificate authority.
+            let cert_key = CertKey::new(vec![cert, ca_cert.clone()], key);
```

Comment:
This line still strikes me as odd.... The client shouldnt' need to send the ca as the server should come pre-populated with the ca. Now, If there was an *intermediate* ca, then absolutely, this would go here in this chain.  As you can see in line 54, 55 of `certificate_callback.rs` you are adding the same ca.pem

```
        // add the CA certificate to the SSL context
        let ca_cert = boring::x509::X509::from_pem(CA_PEM)
```

Reply:
Like, it you removed the equality check below where the Ca checks' it's public key against itself (see below) would the `ca_cert` still need to be included here?
```
// On the client side, the chain includes the leaf certificate, but on the server side it does
        // not. Fun!
        let client_ca_cert = match ssl.peer_cert_chain() {
            Some(chain) => chain.get(0).unwrap(),
            None => {
                log::error!("Failed to get client certificate chain");
                return Err(SslVerifyError::Invalid(SslAlert::NO_CERTIFICATE));
            }
        };
```

Reply:
Makes sense to not include it and instead use what the server has